waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+
+(defn- scale-service-to
+  [api-server-url http-client service instances']
+  (log/info "Scaling to " instances' service)
+  (let [request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (:namespace service)
+                         "/replicasets/"
+                         (:app-name service))
+        instances (:task-count service)]
+    (patch-object-replicas http-client request-url instances instances')))
+
+(defn- scale-service-by
+  [api-server-url http-client service instances-delta]
+  (log/info "Scaling incrementally by " instances-delta service)
+  (let [request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (:namespace service)
+                         "/replicasets/"
+                         (:app-name service))
+        instances (:task-count service)
+        instances' (+ instances instances-delta)]
+    (patch-object-replicas http-client request-url instances instances')))
+
+(defn- kill-service-instance
+  [api-server-url http-client {:keys [id namespace pod-name service-id] :as instance} service]
+  (let [pod-url (str api-server-url
+                     "/api/v1/namespaces/"
+                     namespace
+                     "/pods/"
+                     pod-name)
+        base-body {:kind "DeleteOptions", :apiVersion "v1"}
+        term-json (-> base-body (assoc :gracePeriodSeconds 300) (as-json))
+        kill-json (-> base-body (assoc :gracePeriodSeconds 0) (as-json))
+        make-kill-response (fn [killed? message status]
+                             {:instance-id id, :killed? killed?,
+                              :message message, :service-id service-id, :status status})]
+    ; request termination of the instance
+    (api-request http-client pod-url :request-method :delete :body term-json)
+    ; scale down the replicaset
+    (scale-service-by api-server-url http-client service -1)
+    ; force-kill the instance (should still be terminating)
+    (api-request http-client pod-url :request-method :delete :body kill-json)
+    ; report back that the instance was killed
+    (scheduler/process-instance-killed! instance)
+    (make-kill-response true "pod was killed" 200)))
+
+(defn- service-spec
+  [service-id service-description service-id->password-fn]
+  (let [spec-file-path "./specs/k8s-default-pod.edn" ; TODO - allow user to provide this via the config file
+        {:strs [backend-proto cmd cpus #_disk grace-period-secs health-check-interval-secs
+                health-check-max-consecutive-failures mem ports min-instances
+                #_restart-backoff-factor run-as-user]} service-description
+        home-path (str "/home/" run-as-user)
+        common-env (scheduler/environment service-id service-description
+                                          service-id->password-fn home-path)
+        port0 8080
```

Comment:
Should this value be configurable?

Reply:
I don't see any reason why we couldn't put this in the Waiter config file, and then default to `8080` if it's not explicitly overridden.